src/app/server/CommissioningWindowManager.cpp

```diff
@@ -0,0 +1,277 @@
+#endif // CONFIG_NETWORK_LAYER_BLE
+
+    if (resetFabrics == ResetFabrics::kYes)
+    {
+        mServer->GetFabricTable().DeleteAllFabrics();
+        // Only resetting gNextAvailableFabricIndex at reboot otherwise previously paired device with fabricID 0
+        // can continue sending messages to accessory as next available fabric will also be 0.
+        // This logic is not up to spec, will be implemented up to spec once AddOptCert is implemented.
+        mServer->GetFabricTable().Reset();
+    }
+
+    uint16_t keyID = 0;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    uint32_t pinCode;
+
+    SuccessOrExit(err = DeviceLayer::ConfigurationMgr().GetSetupPinCode(pinCode));
+    SuccessOrExit(err = PrepareCommissioningWindow(commissioningTimeoutSeconds, keyID));
+    SuccessOrExit(err = mPairingSession.WaitForPairing(
+                      pinCode, kSpake2p_Iteration_Count,
+                      ByteSpan(reinterpret_cast<const uint8_t *>(kSpake2pKeyExchangeSalt), strlen(kSpake2pKeyExchangeSalt)), keyID,
+                      this));
+
+    // reset all advertising, indicating we are in commissioningMode
+    // and we were put into this state via a command for additional commissioning
+    app::MdnsServer::Instance().StartServer(Mdns::CommissioningMode::kEnabledBasic);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        Cleanup();
+    }
+    return err;
+}
+
+CHIP_ERROR CommissioningWindowManager::OpenEnhancedCommissioningWindow(uint16_t commissioningTimeoutSeconds, uint16_t discriminator,
+                                                                       PASEVerifier & verifier, uint32_t iterations, ByteSpan salt,
+                                                                       uint16_t passcodeID)
+{
+#if CONFIG_NETWORK_LAYER_BLE
+    // TODO: Don't use BLE for commissioning additional fabrics on a device
+    SetBLE(true);
+#else
+    SetBLE(false);
+#endif
+
+    uint16_t keyID = 0;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    SuccessOrExit(err = SetTemporaryDiscriminator(discriminator));
+    SuccessOrExit(err = PrepareCommissioningWindow(commissioningTimeoutSeconds, keyID));
+    SuccessOrExit(err = mPairingSession.WaitForPairing(verifier, iterations, salt, passcodeID, keyID, this));
+
+    // reset all advertising, indicating we are in commissioningMode
+    // and we were put into this state via a command for additional commissioning
+    app::MdnsServer::Instance().StartServer(Mdns::CommissioningMode::kEnabledEnhanced);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        Cleanup();
+    }
+    return err;
+}
+
+void CommissioningWindowManager::CloseCommissioningWindow()
+{
+    if (mCommissioningWindowOpen)
+    {
+        ChipLogProgress(AppServer, "Closing pairing window");
+        Cleanup();
+    }
+}
+
+CHIP_ERROR CommissioningWindowManager::StartAdvertisement()
+{
+    if (mIsBLE)
+    {
+        ReturnErrorOnFailure(chip::DeviceLayer::ConnectivityMgr().SetBLEAdvertisingEnabled(true));
+    }
```

Comment:
Shouldn't this call to start mDNS commissionable node advertisement for non-BLE mode?

Also, are BLE vs. IP commissioning actually mutually exclusive?

Reply:
This part of the code is actually just enabling the BLE beacon, and is not manipulating the mDNS advertisements. IP commissioning can still be performed even though BLE beacons have been enabled. So IP and BLE are not mutually exclusive.

The goal is to remove BLE commissioning for additional fabric provisioning. Reference TODO: https://github.com/project-chip/connectedhomeip/pull/9709#discussion_r709300501